cmd/collector queued-exporter:jaeger

Zipkin can produce spans without the corresponding proto Node, Jaeger requires a non-nil Process element. In such cases the collector crashes with a nil exception (see below).

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x1390547]

goroutine 38 [running]:
github.com/jaegertracing/jaeger/thrift-gen/jaeger.(*Process).writeField1(0x0, 0x1ccf4a0, 0xc0000e2480, 0x0, 0x0)
        /Users/pj/go/pkg/mod/github.com/jaegertracing/jaeger@v1.8.2/thrift-gen/jaeger/ttypes.go:1541 +0x97
github.com/jaegertracing/jaeger/thrift-gen/jaeger.(*Process).Write(0x0, 0x1ccf4a0, 0xc0000e2480, 0xc00001790c, 0x0)
        /Users/pj/go/pkg/mod/github.com/jaegertracing/jaeger@v1.8.2/thrift-gen/jaeger/ttypes.go:1522 +0x82
github.com/jaegertracing/jaeger/thrift-gen/jaeger.(*Batch).writeField1(0xc00049e220, 0x1ccf4a0, 0xc0000e2480, 0x0, 0x0)
        /Users/pj/go/pkg/mod/github.com/jaegertracing/jaeger@v1.8.2/thrift-gen/jaeger/ttypes.go:1707 +0xae
github.com/jaegertracing/jaeger/thrift-gen/jaeger.(*Batch).Write(0xc00049e220, 0x1ccf4a0, 0xc0000e2480, 0x2c95190, 0xc000076810)
        /Users/pj/go/pkg/mod/github.com/jaegertracing/jaeger@v1.8.2/thrift-gen/jaeger/ttypes.go:1688 +0x82
github.com/census-instrumentation/opencensus-service/cmd/occollector/app/sender.serializeThrift(0x1cb3540, 0xc00049e220, 0xc0000ff6b8, 0x3, 0x3)
        /Users/pj/go/src/github.com/census-instrumentation/opencensus-service/cmd/occollector/app/sender/jaeger_thrift_http_sender.go:127 +0x1a4
github.com/census-instrumentation/opencensus-service/cmd/occollector/app/sender.(*JaegerThriftHTTPSender).ProcessSpans(0xc00018b3b0, 0xc0004aea00, 0x1b5d72b, 0x6, 0x3, 0xc0005421e0, 0x3)
        /Users/pj/go/src/github.com/census-instrumentation/opencensus-service/cmd/occollector/app/sender/jaeger_thrift_http_sender.go:100 +0x1fa
github.com/census-instrumentation/opencensus-service/internal/collector/processor.(*queuedSpanProcessor).processItemFromQueue(0xc0001629c0, 0xc0004ef1a0)
        /Users/pj/go/src/github.com/census-instrumentation/opencensus-service/internal/collector/processor/queued_processor.go:134 +0x9a
github.com/census-instrumentation/opencensus-service/internal/collector/processor.NewQueuedSpanProcessor.func1(0x1970de0, 0xc0004ef1a0)
        /Users/pj/go/src/github.com/census-instrumentation/opencensus-service/internal/collector/processor/queued_processor.go:60 +0x45
github.com/jaegertracing/jaeger/pkg/queue.(*BoundedQueue).StartConsumers.func1(0xc00017a8f0, 0xc000180900, 0xc00017eed0)
        /Users/pj/go/pkg/mod/github.com/jaegertracing/jaeger@v1.8.2/pkg/queue/bounded_queue.go:65 +0xa7
created by github.com/jaegertracing/jaeger/pkg/queue.(*BoundedQueue).StartConsumers
        /Users/pj/go/pkg/mod/github.com/jaegertracing/jaeger@v1.8.2/pkg/queue/bounded_queue.go:58 +0xae
exit status 2
```

--- a/translator/trace/protospan_to_jaegerthrift.go
+++ b/translator/trace/protospan_to_jaegerthrift.go
@@ -37,6 +37,10 @@ var (
 	errWrongLenID      = errors.New("ID does not have 8 bytes")
 )
 
+var (
+	unknownProcess = &jaeger.Process{ServiceName: "unknown-service-name"}
+)
+
 // OCProtoToJaegerThrift translates OpenCensus trace data into the Jaeger Thrift format.
 func OCProtoToJaegerThrift(ocBatch *agenttracepb.ExportTraceServiceRequest) (*jaeger.Batch, error) {
 	if ocBatch == nil {
@@ -58,7 +62,8 @@ func OCProtoToJaegerThrift(ocBatch *agenttracepb.ExportTraceServiceRequest) (*ja
 
 func ocNodeToJaegerProcess(node *commonpb.Node) *jaeger.Process {
 	if node == nil {
-		return nil
+		// Jaeger requires a non-nil Process
+		return unknownProcess
 	}
 
 	var jTags []*jaeger.Tag

--- a/translator/trace/protospan_to_jaegerthrift_test.go
+++ b/translator/trace/protospan_to_jaegerthrift_test.go
@@ -112,6 +112,27 @@ func TestInvalidOCProtoIDs(t *testing.T) {
 	}
 }
 
+func TestNilOCProtoNode(t *testing.T) {
+	nilNodeBatch := &agenttracepb.ExportTraceServiceRequest{
+		Spans: []*tracepb.Span{
+			{
+				TraceId: []byte("0123456789abcdef"),
+				SpanId:  []byte("01234567"),
+			},
+		},
+	}
+	got, err := OCProtoToJaegerThrift(nilNodeBatch)
+	if err != nil {
+		t.Fatalf("Failed to translate OC batch to Jaeger Thrift: %v", err)
+	}
+	if got.Process == nil {
+		t.Fatalf("Jaeger requires a non-nil Process field")
+	}
+	if got.Process != unknownProcess {
+		t.Fatalf("got unexpected Jaeger Process field")
+	}
+}
+
 func TestOCProtoToJaegerThrift(t *testing.T) {
 	const numOfFiles = 2
 	for i := 0; i < numOfFiles; i++ {

